sql: fix TestAuthenticationAndHBARules stress tests SSL issue

souravcrl opened the pull request:
sql: fix TestAuthenticationAndHBARules stress tests SSL issue

informs #131532
informs #131110
informs #130253
informs #127745
Epic CRDB-41958

TestAuthenticationAndHBARules fails for `special_cases`, `hba_default_equivalence`, `empty_hba` data driven tests for secure mode. The failures occur when root user is trying to authenticate with cert-password auth method and `sslmode` is set to `verify-ca` with `sslcert` being empty. The expected behavior is root authentication defaults to password method and fails as no password is set for root, but instead we get:
```
pq: SSL is not enabled on the server
```
Since the failures are there only under stress, it might be because db server shutdown or paused before responding to request for upgrade connection to SSL from lib/pq client from here
https://github.com/lib/pq/blob/3d613208bca2e74f2a20e04126ed30bcb5c4cc27/conn.go#L1116-L1130. Retrying connection establishment when this specific error is obtained might fix the problem as this logic seems faulty(it checks for absence of 'S' in server response whereas the correct check should be for 'N' in response).

----

pgwire: add test build logs for upgrade secure conn

Additional test build logs are added to verify the step for which
`maybeUpgradeToSecureConn` fails.

Release note: None

souravcrl committed:
sql: fix TestAuthenticationAndHBARules stress tests SSL issue

informs #131532
informs #131110
informs #130253
informs #127745
Epic CRDB-41958

TestAuthenticationAndHBARules fails for `special_cases`,
`hba_default_equivalence`, `empty_hba` data driven tests for secure mode. The
failures occur when root user is trying to authenticate with cert-password auth
method and `sslmode` is set to `verify-ca` with `sslcert` being empty. The
expected behavior is root authentication defaults to password method and fails
as no password is set for root, but instead we get:
```
SSL is not enabled on the server
```
Since the failures are there only under stress, it might be because db server
shutdown or paused before responding to request for upgrade connection to SSL
from lib/pq client from here
https://github.com/lib/pq/blob/3d613208bca2e74f2a20e04126ed30bcb5c4cc27/conn.go#L1116-L1130.
Retrying connection establishment when this specific error is obtained might fix
the problem as this logic seems faulty(it checks for absence of 'S' in server
response whereas the correct check should be for 'N' in response).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/132729)
<!-- Reviewable:end -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132729)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pritesh-lahoti and @souravcrl)

___
*[`pkg/sql/pgwire/auth_test.go` line 575 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132729#-O9R7NZ0BhnhwGu8CyKz:-O9R7NZ0BhnhwGu8CyL-:bfth4ya) ([raw file](https://github.com/cockroachdb/cockroach/blob/69b511ccddc8afef0c6ceefabd573324ae22f38c/pkg/sql/pgwire/auth_test.go#L575)):*
> ```Go
> 					}
> 					if err != nil {
> 						if !strings.Contains(err.Error(), "SSL is not enabled on the server") {
> ```

let's use `errors.Is(err, pq.ErrSSLNotSupported)` rather than a string match

https://github.com/lib/pq/blob/3d613208bca2e74f2a20e04126ed30bcb5c4cc27/conn.go#L34
___
*[`pkg/sql/pgwire/auth_test.go` line 579 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132729#-O9R7XbI-NSt2-MaWIp7:-O9R7XbI-NSt2-MaWIp8:b-6z587c) ([raw file](https://github.com/cockroachdb/cockroach/blob/69b511ccddc8afef0c6ceefabd573324ae22f38c/pkg/sql/pgwire/auth_test.go#L579)):*
> ```Go
> 						}
> 						// Retry: server might fail to respond to upgrade conn under stress
> 						dbSQL, err = gosql.Open("postgres", dsn)
> ```

hmm i'd like to understand a little bit more about how this might help before merging. let me dive into this a little bit first.


<!-- Sent from Reviewable.io -->

souravcrl reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132729)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pritesh-lahoti and @rafiss)

___
*[`pkg/sql/pgwire/auth_test.go` line 575 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132729#-O9R7NZ0BhnhwGu8CyKz:-O9TnVujBv-C_Bx7X52v:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/69b511ccddc8afef0c6ceefabd573324ae22f38c/pkg/sql/pgwire/auth_test.go#L575)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

let's use `errors.Is(err, pq.ErrSSLNotSupported)` rather than a string match

https://github.com/lib/pq/blob/3d613208bca2e74f2a20e04126ed30bcb5c4cc27/conn.go#L34
</blockquote></details>

done
___
*[`pkg/sql/pgwire/auth_test.go` line 579 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132729#-O9R7XbI-NSt2-MaWIp7:-O9TnYlq-5oCrtjiBrqA:b-ftzy1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/69b511ccddc8afef0c6ceefabd573324ae22f38c/pkg/sql/pgwire/auth_test.go#L579)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hmm i'd like to understand a little bit more about how this might help before merging. let me dive into this a little bit first.
</blockquote></details>

have taken a note of this. will await your reply.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132729)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pritesh-lahoti and @souravcrl)

___
*[`pkg/sql/pgwire/auth_test.go` line 579 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132729#-O9R7XbI-NSt2-MaWIp7:-O9VdNzZ1dflrnzJwwfq:b-5rw38m) ([raw file](https://github.com/cockroachdb/cockroach/blob/69b511ccddc8afef0c6ceefabd573324ae22f38c/pkg/sql/pgwire/auth_test.go#L579)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

have taken a note of this. will await your reply.
</blockquote></details>

i think we should try to gather more info on the failure before going with a blind retry. (unless, have you already done some of this investigation?)

i think a patch like this will help us see exactly which step of upgrading the secure conn is not working:
```
diff --git a/pkg/sql/pgwire/pre_serve.go b/pkg/sql/pgwire/pre_serve.go
index 66d1d81f86c..d0bca6747c1 100644
--- a/pkg/sql/pgwire/pre_serve.go
+++ b/pkg/sql/pgwire/pre_serve.go
@@ -23,6 +23,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgwirebase"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgwirecancel"
+	"github.com/cockroachdb/cockroach/pkg/util/buildutil"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/metric"
 	"github.com/cockroachdb/cockroach/pkg/util/mon"
@@ -446,6 +447,10 @@ func (s *PreServeConnHandler) maybeUpgradeToSecureConn(
 		// Secure mode: disallow if TCP and the user did not opt into
 		// non-TLS SQL conns.
 		if !s.cfg.AcceptSQLWithoutTLS && connType != hba.ConnLocal && connType != hba.ConnInternalLoopback {
+			// Extra logs under test to debug TestAuthenticationAndHBARules.
+			if buildutil.CrdbTestBuild {
+				log.Warningf(ctx, "client cannot connect since version=%d AcceptSQLWithoutTLS=false and connType=%s", version, connType)
+			}
 			clientErr = pgerror.New(pgcode.ProtocolViolation, ErrSSLRequired)
 		}
 		return
@@ -458,6 +463,10 @@ func (s *PreServeConnHandler) maybeUpgradeToSecureConn(
 		// we don't want it.
 		clientErr = pgerror.New(pgcode.ProtocolViolation,
 			"cannot use SSL/TLS over local connections")
+		// Extra logs under test to debug TestAuthenticationAndHBARules.
+		if buildutil.CrdbTestBuild {
+			log.Warningf(ctx, "client cannot connect since version=%d and connType=%s", version, connType)
+		}
 		return
 	}
 
@@ -479,6 +488,10 @@ func (s *PreServeConnHandler) maybeUpgradeToSecureConn(
 	if tlsConfig == nil {
 		// We don't have a TLS configuration available, so we can't honor
 		// the client's request.
+		// Extra logs under test to debug TestAuthenticationAndHBARules.
+		if buildutil.CrdbTestBuild {
+			log.Infof(ctx, "sending sslUnsupported message to client")
+		}
 		n, serverErr = conn.Write(sslUnsupported)
 		if serverErr != nil {
 			return
```

we can merge it, and then next time it fails we should have a helpful clue. what do you think about this?


<!-- Sent from Reviewable.io -->

souravcrl committed:
pgwire: add test build logs for upgrade secure conn

Epic CRDB-41958

Additional test build logs are added to verify the step for which
`maybeUpgradeToSecureConn` fails for `TestAuthenticationAndHBARules`.

Release note: None

souravcrl reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132729)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pritesh-lahoti and @rafiss)

___
*[`pkg/sql/pgwire/auth_test.go` line 579 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132729#-O9R7XbI-NSt2-MaWIp7:-O9k48J38rLgsjMDpIcc:b-a5n12w) ([raw file](https://github.com/cockroachdb/cockroach/blob/69b511ccddc8afef0c6ceefabd573324ae22f38c/pkg/sql/pgwire/auth_test.go#L579)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think we should try to gather more info on the failure before going with a blind retry. (unless, have you already done some of this investigation?)

i think a patch like this will help us see exactly which step of upgrading the secure conn is not working:
```
diff --git a/pkg/sql/pgwire/pre_serve.go b/pkg/sql/pgwire/pre_serve.go
index 66d1d81f86c..d0bca6747c1 100644
--- a/pkg/sql/pgwire/pre_serve.go
+++ b/pkg/sql/pgwire/pre_serve.go
@@ -23,6 +23,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgwirebase"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgwirecancel"
+	"github.com/cockroachdb/cockroach/pkg/util/buildutil"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/metric"
 	"github.com/cockroachdb/cockroach/pkg/util/mon"
@@ -446,6 +447,10 @@ func (s *PreServeConnHandler) maybeUpgradeToSecureConn(
 		// Secure mode: disallow if TCP and the user did not opt into
 		// non-TLS SQL conns.
 		if !s.cfg.AcceptSQLWithoutTLS && connType != hba.ConnLocal && connType != hba.ConnInternalLoopback {
+			// Extra logs under test to debug TestAuthenticationAndHBARules.
+			if buildutil.CrdbTestBuild {
+				log.Warningf(ctx, "client cannot connect since version=%d AcceptSQLWithoutTLS=false and connType=%s", version, connType)
+			}
 			clientErr = pgerror.New(pgcode.ProtocolViolation, ErrSSLRequired)
 		}
 		return
@@ -458,6 +463,10 @@ func (s *PreServeConnHandler) maybeUpgradeToSecureConn(
 		// we don't want it.
 		clientErr = pgerror.New(pgcode.ProtocolViolation,
 			"cannot use SSL/TLS over local connections")
+		// Extra logs under test to debug TestAuthenticationAndHBARules.
+		if buildutil.CrdbTestBuild {
+			log.Warningf(ctx, "client cannot connect since version=%d and connType=%s", version, connType)
+		}
 		return
 	}
 
@@ -479,6 +488,10 @@ func (s *PreServeConnHandler) maybeUpgradeToSecureConn(
 	if tlsConfig == nil {
 		// We don't have a TLS configuration available, so we can't honor
 		// the client's request.
+		// Extra logs under test to debug TestAuthenticationAndHBARules.
+		if buildutil.CrdbTestBuild {
+			log.Infof(ctx, "sending sslUnsupported message to client")
+		}
 		n, serverErr = conn.Write(sslUnsupported)
 		if serverErr != nil {
 			return
```

we can merge it, and then next time it fails we should have a helpful clue. what do you think about this?
</blockquote></details>

updated the PR with the desired logs. have the backport out https://github.com/cockroachdb/cockroach/pull/133002 and running stress tests.


<!-- Sent from Reviewable.io -->

souravcrl reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132729)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pritesh-lahoti and @rafiss)

___
*[`pkg/sql/pgwire/auth_test.go` line 579 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132729#-O9R7XbI-NSt2-MaWIp7:-O9kDin56ctBmeKgoUxK:b8yvv8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/69b511ccddc8afef0c6ceefabd573324ae22f38c/pkg/sql/pgwire/auth_test.go#L579)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

updated the PR with the desired logs. have the backport out https://github.com/cockroachdb/cockroach/pull/133002 and running stress tests.
</blockquote></details>

running it here: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/17398033?hideProblemsFromDependencies=false&hideTestsFromDependencies=false


<!-- Sent from Reviewable.io -->

souravcrl reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132729)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pritesh-lahoti and @rafiss)

___
*[`pkg/sql/pgwire/auth_test.go` line 579 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132729#-O9R7XbI-NSt2-MaWIp7:-O9my0M7AnydNPv_wZZI:b7ecz2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/69b511ccddc8afef0c6ceefabd573324ae22f38c/pkg/sql/pgwire/auth_test.go#L579)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

running it here: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/17398033?hideProblemsFromDependencies=false&hideTestsFromDependencies=false
</blockquote></details>

the build ran successfully for 11h42m without failures. Should we proceed with the PR?


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132729)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pritesh-lahoti and @souravcrl)

___
*[`pkg/sql/pgwire/auth_test.go` line 579 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132729#-O9R7XbI-NSt2-MaWIp7:-O9pmWsu0sLD4WNSV0MU:b-j96042) ([raw file](https://github.com/cockroachdb/cockroach/blob/69b511ccddc8afef0c6ceefabd573324ae22f38c/pkg/sql/pgwire/auth_test.go#L579)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

the build ran successfully for 11h42m without failures. Should we proceed with the PR?
</blockquote></details>

my thinking was that we would avoid adding this retry logic until we get more information that tells us why it will help.

so let's just merge the logging change, then see what we find in the next failure after that.


<!-- Sent from Reviewable.io -->

souravcrl reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132729)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pritesh-lahoti and @rafiss)

___
*[`pkg/sql/pgwire/auth_test.go` line 579 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132729#-O9R7XbI-NSt2-MaWIp7:-O9s-Ruk9PSWS1N3MjXI:b-kfm36e) ([raw file](https://github.com/cockroachdb/cockroach/blob/69b511ccddc8afef0c6ceefabd573324ae22f38c/pkg/sql/pgwire/auth_test.go#L579)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

my thinking was that we would avoid adding this retry logic until we get more information that tells us why it will help.

so let's just merge the logging change, then see what we find in the next failure after that.
</blockquote></details>

have the test only build logs PR out: https://github.com/cockroachdb/cockroach/pull/133224.
Also have the backports out here:

[release-23.1: pgwire: add test build logs for upgrade secure conn #133225](https://github.com/cockroachdb/cockroach/pull/133225)

[release-23.2: pgwire: add test build logs for upgrade secure conn #133226](https://github.com/cockroachdb/cockroach/pull/133226)

[release-24.1: pgwire: add test build logs for upgrade secure conn #133227](https://github.com/cockroachdb/cockroach/pull/133227)

[release-24.2: pgwire: add test build logs for upgrade secure conn #133228](https://github.com/cockroachdb/cockroach/pull/133228)

[release-24.3: pgwire: add test build logs for upgrade secure conn #133229](https://github.com/cockroachdb/cockroach/pull/133229)


<!-- Sent from Reviewable.io -->
